File: src/Azdo.Tui/Components/StatusBar.cs
using Azdo.Tui.Polling;
using Azdo.Tui.Rendering;

namespace Azdo.Tui.Components;

/// <summary>
/// Footer showing context/default keybindings, org/project, connection state,
/// filter labels, warnings and scroll percentage (≈ <c>components.StatusBar</c>).
/// Mutable, matching the Go pointer-based component.
/// </summary>
public sealed class StatusBar(Styles.Styles styles)
{
    private string _organization = "";
    private string _project = "";
    private ConnectionState _state = ConnectionState.Connecting;
    private string _keybindings = "";
    private double _scrollPercent;
    private bool _showScroll;
    private int _width;
    private string _errorMessage = "";
    private string _filterLabel = "";
    private string _updateMessage = "";
    private string _warningMessage = "";
    private IReadOnlyList<ContextItem> _contextItems = Array.Empty<ContextItem>();
    private string _contextStatus = "";

    public void SetOrganization(string org) => _organization = org;
    public void SetProject(string project) => _project = project;
    public ConnectionState GetState() => _state;
    public void SetState(ConnectionState state) => _state = state;
    public string GetWarningMessage() => _warningMessage;
    public void SetKeybindings(string bindings) => _keybindings = bindings;
    public void SetContextItems(IReadOnlyList<ContextItem> items) => _contextItems = items;
    public void ClearContextItems() { _contextItems = Array.Empty<ContextItem>(); _contextStatus = ""; }
    public void SetContextStatus(string status) => _contextStatus = status;
    public void SetWidth(int width) => _width = width;
    public void SetScrollPercent(double percent) => _scrollPercent = percent;
    public void ShowScrollPercent(bool show) => _showScroll = show;
    public void SetErrorMessage(string message) => _errorMessage = message;
    public void ClearErrorMessage() => _errorMessage = "";
    public void SetFilterLabel(string label) => _filterLabel = label;
    public void ClearFilterLabel() => _filterLabel = "";
    public void SetUpdateMessage(string message) => _updateMessage = message;
    public void SetWarningMessage(string message) => _warningMessage = message;
    public void ClearWarningMessage() => _warningMessage = "";

    public string View()
    {
        int width = _width < 40 ? 80 : _width;
        var sepStyle = Style.New().Foreground(styles.Theme.Border);
        var sep = sepStyle.Render(" │ ");

        var parts = new List<string>();

        if (_errorMessage != "" && _state == ConnectionState.Error)
            parts.Add(Style.New().Foreground(styles.Theme.Error).Bold().Render(_errorMessage));
        else
            parts.Add(RenderKeybindings());

        if (_contextStatus != "")
            parts.Add(Style.New().Foreground(styles.Theme.ForegroundMuted).Italic().Render(_contextStatus));

        if (_warningMessage != "")
            parts.Add(Style.New().Foreground(styles.Theme.Warning).Bold().Render("⚠ " + _warningMessage));

        if (_filterLabel != "")
            parts.Add(Style.New().Foreground(styles.Theme.Background).Background(styles.Theme.Accent).Bold().Padding(0, 1).Render(_filterLabel));

        if (_updateMessage != "")
            parts.Add(Style.New().Foreground(styles.Theme.Warning).Bold().Render(_updateMessage));

        var orgProj = RenderOrgProject();
        if (orgProj != "") parts.Add(orgProj);

        var scroll = RenderScrollPercent();
        if (scroll != "") parts.Add(scroll);

        parts.Add(RenderConnectionState());

        var content = string.Join(sep, parts);
        int boxInnerWidth = Math.Max(20, width - 2);
        return styles.BoxRounded.Width(boxInnerWidth).Render(content);
    }

    private string RenderKeybindings()
    {
        if (_contextItems.Count > 0) return RenderContextKeybindings();
        if (_keybindings != "") return _keybindings;

        var sep = Style.New().Foreground(styles.Theme.Border).Render(" • ");
        return styles.Key.Render("r") + styles.Description.Render(" refresh") + sep +
               styles.Key.Render("↑↓") + styles.Description.Render(" navigate") + sep +
               styles.Key.Render("enter") + styles.Description.Render(" details") + sep +
               styles.Key.Render("esc") + styles.Description.Render(" back") + sep +
               styles.Key.Render("?") + styles.Description.Render(" help") + sep +
               styles.Key.Render("q") + styles.Description.Render(" quit");
    }

    private string RenderContextKeybindings()
    {
        var sep = Style.New().Foreground(styles.Theme.Border).Render(" • ");
        var hasKey = _contextItems.Select(i => i.Key).ToHashSet();
        var parts = _contextItems
            .Select(i => styles.Key.Render(i.Key) + " " + styles.Description.Render(i.Description))
            .ToList();

        foreach (var (key, desc) in new[] { ("esc", "back"), ("?", "help"), ("q", "quit") })
            if (!hasKey.Contains(key))
                parts.Add(styles.Key.Render(key) + " " + styles.Description.Render(desc));

        return string.Join(sep, parts);
    }

    private string RenderOrgProject()
    {
        if (_organization == "" && _project == "") return "";
        var sep = Style.New().Foreground(styles.Theme.Border).Render("/");
        var s = Style.New().Foreground(styles.Theme.Secondary).Bold();
        if (_organization != "" && _project != "") return s.Render(_organization) + sep + s.Render(_project);
        return _organization != "" ? s.Render(_organization) : s.Render(_project);
    }

    private string RenderScrollPercent()
        => !_showScroll ? "" : styles.ScrollInfo.Render($"{_scrollPercent:0}%");

    private string RenderConnectionState() => _state switch
    {
        ConnectionState.Connected => styles.Connected.Render("●"),
        ConnectionState.Connecting => styles.Connecting.Render("◐ connecting"),
        ConnectionState.Disconnected => styles.Disconnected.Render("○ disconnected"),
        ConnectionState.Error => styles.ConnError.Render("✗ error"),
        _ => styles.Disconnected.Render($"? {_state}"),
    };
}

File: reference/internal/ui/components/statusbar.go
// Package components provides reusable UI components for the TUI.
package components

import (
	"fmt"
	"strings"

	"github.com/Elpulgo/azdo/internal/polling"
	"github.com/Elpulgo/azdo/internal/ui/styles"
	tea "github.com/charmbracelet/bubbletea"
	"github.com/charmbracelet/lipgloss"
)

// StatusBar is a component that displays keybindings, org/project info,
// and connection state at the bottom of the screen.
type StatusBar struct {
	styles         *styles.Styles
	organization   string
	project        string
	state          polling.ConnectionState
	keybindings    string
	scrollPercent  float64
	showScroll     bool
	width          int
	errorMessage   string
	filterLabel    string
	updateMessage  string
	warningMessage string
	contextItems   []ContextItem
	contextStatus  string
}

// NewStatusBar creates a new StatusBar with default values.
func NewStatusBar(s *styles.Styles) *StatusBar {
	return &StatusBar{
		styles:      s,
		state:       polling.StateConnecting,
		keybindings: "",
	}
}

// SetOrganization sets the organization name to display.
func (s *StatusBar) SetOrganization(org string) {
	s.organization = org
}

// SetProject sets the project name to display.
func (s *StatusBar) SetProject(project string) {
	s.project = project
}

// GetState returns the current connection state.
func (s *StatusBar) GetState() polling.ConnectionState {
	return s.state
}

// SetState sets the connection state.
func (s *StatusBar) SetState(state polling.ConnectionState) {
	s.state = state
}

// GetWarningMessage returns the current warning message.
func (s *StatusBar) GetWarningMessage() string {
	return s.warningMessage
}

// SetKeybindings sets the keybindings to display.
func (s *StatusBar) SetKeybindings(bindings string) {
	s.keybindings = bindings
}

// SetContextItems sets context-specific keybindings that replace the default
// keybindings in the footer. Base shortcuts (esc back, ? help, q quit) are
// automatically appended with deduplication.
func (s *StatusBar) SetContextItems(items []ContextItem) {
	s.contextItems = items
}

// ClearContextItems removes context-specific keybindings, restoring defaults.
func (s *StatusBar) ClearContextItems() {
	s.contextItems = nil
	s.contextStatus = ""
}

// SetContextStatus sets an italic status message displayed alongside context items.
func (s *StatusBar) SetContextStatus(status string) {
	s.contextStatus = status
}

// SetWidth sets the width of the status bar.
func (s *StatusBar) SetWidth(width int) {
	s.width = width
}

// SetScrollPercent sets the scroll percentage (0-100).
func (s *StatusBar) SetScrollPercent(percent float64) {
	s.scrollPercent = percent
}

// ShowScrollPercent enables or disables showing the scroll percentage.
func (s *StatusBar) ShowScrollPercent(show bool) {
	s.showScroll = show
}

// SetErrorMessage sets the error message to display.
func (s *StatusBar) SetErrorMessage(message string) {
	s.errorMessage = message
}

// ClearErrorMessage clears the error message.
func (s *StatusBar) ClearErrorMessage() {
	s.errorMessage = ""
}

// SetFilterLabel sets a filter indicator label to display in the status bar.
func (s *StatusBar) SetFilterLabel(label string) {
	s.filterLabel = label
}

// ClearFilterLabel removes the filter indicator label.
func (s *StatusBar) ClearFilterLabel() {
	s.filterLabel = ""
}

// SetUpdateMessage sets the update notification message.
func (s *StatusBar) SetUpdateMessage(message string) {
	s.updateMessage = message
}

// SetWarningMessage sets a persistent warning message that displays regardless of connection state.
func (s *StatusBar) SetWarningMessage(message string) {
	s.warningMessage = message
}

// ClearWarningMessage clears the persistent warning message.
func (s *StatusBar) ClearWarningMessage() {
	s.warningMessage = ""
}

// Init implements tea.Model (no initialization needed).
func (s *StatusBar) Init() tea.Cmd {
	return nil
}

// Update implements tea.Model (status bar doesn't handle messages).
func (s *StatusBar) Update(msg tea.Msg) (*StatusBar, tea.Cmd) {
	return s, nil
}

// View renders the status bar as a full-width footer with box border.
func (s *StatusBar) View() string {
	// Use terminal width or default
	width := s.width
	if width < 40 {
		width = 80
	}

	// Build separator style
	sepStyle := lipgloss.NewStyle().
		Foreground(lipgloss.Color(s.styles.Theme.Border))

	sep := sepStyle.Render(" │ ")

	parts := []string{}

	// If there's an error message and state is error, show it prominently
	if s.errorMessage != "" && s.state == polling.StateError {
		errorStyle := lipgloss.NewStyle().
			Foreground(lipgloss.Color(s.styles.Theme.Error)).
			Bold(true)
		parts = append(parts, errorStyle.Render(s.errorMessage))
	} else {
		parts = append(parts, s.renderKeybindings())
	}

	if s.contextStatus != "" {
		statusStyle := lipgloss.NewStyle().
			Foreground(lipgloss.Color(s.styles.Theme.ForegroundMuted)).
			Italic(true)
		parts = append(parts, statusStyle.Render(s.contextStatus))
	}

	if s.warningMessage != "" {
		warningStyle := lipgloss.NewStyle().
			Foreground(lipgloss.Color(s.styles.Theme.Warning)).
			Bold(true)
		parts = append(parts, warningStyle.Render("⚠ "+s.warningMessage))
	}

	if s.filterLabel != "" {
		filterStyle := lipgloss.NewStyle().
			Foreground(lipgloss.Color(s.styles.Theme.Background)).
			Background(lipgloss.Color(s.styles.Theme.Accent)).
			Bold(true).
			Padding(0, 1)
		parts = append(parts, filterStyle.Render(s.filterLabel))
	}

	if s.updateMessage != "" {
		updateStyle := lipgloss.NewStyle().
			Foreground(lipgloss.Color(s.styles.Theme.Warning)).
			Bold(true)
		parts = append(parts, updateStyle.Render(s.updateMessage))
	}

	if orgProj := s.renderOrgProject(); orgProj != "" {
		parts = append(parts, orgProj)
	}

	if scrollPercent := s.renderScrollPercent(); scrollPercent != "" {
		parts = append(parts, scrollPercent)
	}

	parts = append(parts, s.renderConnectionState())

	// Join with separators, left-aligned
	content := strings.Join(parts, sep)

	// Calculate box inner width (subtract 2 for border sides)
	boxInnerWidth := width - 2
	if boxInnerWidth < 20 {
		boxInnerWidth = 20
	}

	return s.styles.BoxRounded.Width(boxInnerWidth).Render(content)
}

// renderKeybindings renders the keybindings section.
func (s *StatusBar) renderKeybindings() string {
	if len(s.contextItems) > 0 {
		return s.renderContextKeybindings()
	}

	if s.keybindings != "" {
		return s.keybindings
	}

	// Build styles from theme
	sepStyle := lipgloss.NewStyle().
		Foreground(lipgloss.Color(s.styles.Theme.Border))

	// Default keybindings with styled keys
	sep := sepStyle.Render(" • ")
	return s.styles.Key.Render("r") + s.styles.Description.Render(" refresh") + sep +
		s.styles.Key.Render("↑↓") + s.styles.Description.Render(" navigate") + sep +
		s.styles.Key.Render("enter") + s.styles.Description.Render(" details") + sep +
		s.styles.Key.Render("esc") + s.styles.Description.Render(" back") + sep +
		s.styles.Key.Render("?") + s.styles.Description.Render(" help") + sep +
		s.styles.Key.Render("q") + s.styles.Description.Render(" quit")
}

// renderContextKeybindings renders context-specific keybindings with base
// shortcuts (esc back, ? help, q quit) appended, deduplicating any that
// already exist in the context items.
func (s *StatusBar) renderContextKeybindings() string {
	sepStyle := lipgloss.NewStyle().
		Foreground(lipgloss.Color(s.styles.Theme.Border))
	sep := sepStyle.Render(" • ")

	// Track which base shortcut keys are already in context items
	hasKey := make(map[string]bool)
	for _, item := range s.contextItems {
		hasKey[item.Key] = true
	}

	// Render context items
	var parts []string
	for _, item := range s.contextItems {
		parts = append(parts, s.styles.Key.Render(item.Key)+" "+s.styles.Description.Render(item.Description))
	}

	// Append base shortcuts with deduplication
	baseShortcuts := []ContextItem{
		{Key: "esc", Description: "back"},
		{Key: "?", Description: "help"},
		{Key: "q", Description: "quit"},
	}
	for _, base := range baseShortcuts {
		if !hasKey[base.Key] {
			parts = append(parts, s.styles.Key.Render(base.Key)+" "+s.styles.Description.Render(base.Description))
		}
	}

	return strings.Join(parts, sep)
}

// renderOrgProject renders the organization and project section.
func (s *StatusBar) renderOrgProject() string {
	if s.organization == "" && s.project == "" {
		return ""
	}

	sepStyle := lipgloss.NewStyle().
		Foreground(lipgloss.Color(s.styles.Theme.Border))

	orgProjectStyle := lipgloss.NewStyle().
		Foreground(lipgloss.Color(s.styles.Theme.Secondary)).
		Bold(true)

	sep := sepStyle.Render("/")

	if s.organization != "" && s.project != "" {
		return orgProjectStyle.Render(s.organization) + sep + orgProjectStyle.Render(s.project)
	}

	if s.organization != "" {
		return orgProjectStyle.Render(s.organization)
	}

	return orgProjectStyle.Render(s.project)
}

// renderScrollPercent renders the scroll percentage indicator.
func (s *StatusBar) renderScrollPercent() string {
	if !s.showScroll {
		return ""
	}
	return s.styles.ScrollInfo.Render(fmt.Sprintf("%.0f%%", s.scrollPercent))
}

// renderConnectionState renders the connection state indicator.
func (s *StatusBar) renderConnectionState() string {
	switch s.state {
	case polling.StateConnected:
		return s.styles.Connected.Render("●")
	case polling.StateConnecting:
		return s.styles.Connecting.Render("◐ connecting")
	case polling.StateDisconnected:
		return s.styles.Disconnected.Render("○ disconnected")
	case polling.StateError:
		return s.styles.ConnError.Render("✗ error")
	default:
		return s.styles.Disconnected.Render(fmt.Sprintf("? %s", s.state))
	}
}
